### PR TITLE
Check override + implement method signature #30

### PR DESCRIPTION
**Pull request by soywiz**

### CODE REVIEW

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java [hunk 1]
```diff
@@ -44,26 +47,35 @@ public void annotate(PsiElement element, AnnotationHolder holder) {
...
+    }
+    else if (element instanceof HaxeMethod) {
       MethodChecker.check((HaxeMethod)element, holder);
-    } else if (element instanceof HaxeClass) {
+    }
+    else if (element instanceof HaxeClass) {
       ClassChecker.check((HaxeClass)element, holder);
-    } if (element instanceof HaxeType) {
+    }
+    if (element instanceof HaxeType) {
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: The clauses above all have 'else if' while this line and the one below (line #60) only have an 'if' statement.  Either construct should work because the types are all independent.  The else clause is probably better, though.  At least we should make them consistent.

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java [hunk 2]
```diff
@@ -413,16 +442,83 @@ public void run() {
...
+      }
+    }
+  }
+
+  static public void checkMethodsSignatureCompatibility(
+    @NotNull final HaxeMethodModel currentMethod,
+    @NotNull final HaxeMethodModel parentMethod,
+    final AnnotationHolder holder
+  ) {
+    final HaxeDocumentModel document = currentMethod.getDocument();
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: This routine will fail when applied to optional arguments.  http://haxe.org/manual/types-function-optional-arguments.html

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: I think that it's OK to address optional arguments in a follow-on PR.

File: testSrc/com/intellij/plugins/haxe/ide/HaxeSemanticAnnotatorTest.java [hunk 3]
```diff
@@ -154,4 +154,16 @@ public void testPropertiesSimpleCheck() throws Exception {
   public void testPropertyAllowNonConstantInitialization() throws Exception {
     doTestNoFixWithWarnings();
   }
+
+  public void testOverrideSignature() throws Exception {
+    doTest("Remove argument");
+  }
+
```
Comment (EBatTiVo) [inline testSrc/com/intellij/plugins/haxe/ide/HaxeSemanticAnnotatorTest.java]: How about a couple of tests to ensure that the comma detection and deleting in remove() works as expected?

File: src/common/com/intellij/plugins/haxe/model/HaxeParameterModel.java [hunk 4]
```diff
@@ -104,4 +109,20 @@ public String getPresentableText() {
...
+    PsiElement psi = getPsi();
+    if (psi != null) {
+      PsiElement prePsi = UsefulPsiTreeUtil.getPrevSiblingSkipWhiteSpaces(psi, true);
+      PsiElement nextPsi = UsefulPsiTreeUtil.getNextSiblingNoSpaces(psi);
+      TextRange range = psi.getTextRange();
+
+      if (prePsi != null && prePsi.getText().equals(",")) {
+        range = range.union(prePsi.getTextRange());
+      } else if (nextPsi != null && nextPsi.getText().equals(",")) {
+        range = range.union(nextPsi.getTextRange());
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/model/HaxeParameterModel.java]: Looks like it might leave two commas in place in the case: (existing, toRemove, existing)

**Comment by EBatTiVo** [comment]
Nice, Carlos.  I'm not trying to be picky; just looking for bugs.

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java (hunk 1 above)
Comment (soywiz) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: Didn't notice. Probably I separated it with some empty lines, but when reformatted they were combined as if it was a single if...else chain. I have examined it, and I think they could be all if else. In other case I could always separate it in three methods but I think it is not the case.

I highly agree with you that it should be consistent. I will fix it in next commit.

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java (hunk 2 above)
Comment (soywiz) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: I have added a test and it works. That's because the ? character is part of the Parameter subtree.

capture

File: src/common/com/intellij/plugins/haxe/model/HaxeParameterModel.java (hunk 4 above)
Comment (soywiz) [inline src/common/com/intellij/plugins/haxe/model/HaxeParameterModel.java]: I have added a test that remove a parameter between two other parameters.
```a, ~~b~~, c```

I'm not sure if I understood right, but since it is an else if instead of an if, it would just remove one comma instead of two in the case there are commands, and would work fine.

I have added two more test cases (and fixed a small whitespace issue):
remove a parameter between two other parameters, and remove the first parameter.

**Comment by soywiz** [comment]
@EBatTiVo Don't worry about code reviews. I'm not offended or anything. And they are good for improving or clarifying. I always appreciate your comments.

Ok. So I have handled the requested stuff (I think). Now it has more coverage than before (and I have also fixed a issue with whitespaces) and I have fixed too the else if issue.

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java (hunk 2 above)
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: What I meant was that the check will not work correctly if the method being checked against (parent) contains optional arguments that are not included in the call being checked.  For example:

```haxe
  function doSomething(int:a, ?String:b, int:c) {}

  function doingSomethingelse() {
    doSomething(1,3);   //checkMethodSignatureCompatibilty will add "Non-matching arity expeced..." error.
  }
```
The error will be incorrect because the number of parameters check does not take into account that some of the parameters are optional.

Then the loop afterward, does not take into account that a type may match a later argument when the current argument is optional.

A solution may be to generate the list of correct signatures from the parent, and then see if the current signature matches any of them.  Then, if not find the closest matching one (the one that would mark the fewest errors?) and mark errors using that.

Comment (soywiz) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: checkMethodsSignatureCompatibility doesn't check function calls, but function definitions (to check overrides and implementations). Not sure if haxe allows you to skip optional arguments. I didn't know that it allowed to have optional arguments in any part of the signature, and call methods as you put in the example, that's a bit odd.

Method calls checks are not in SemanticAnalyzer class but, in order to avoid duplicated code, they are part of the expression evaluator which accepts a context to emit errors.

But i would treat oddities like that one in other pull request, providing false-positive and missing errors as tests to avoid regressions.

Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: The link to the manual I put in the first comment talks about how optional arguments work -- and that they can appear before other non-optional arguments.  A little empirical testing (http://try.haxe.org/#651C8) shows excluding optional arguments to be allowed only for function calls.  For declarations and definitions (overrides and implementations), two functions in the class hierarchy can't use the same name with different signatures, and override signatures have to be identical, including the optional fields.

That all means this implementation is correct for checking declarations.  As it is, the function can't be used to check call parameters, but that's OK since it's not in the models, but in the annotator itself.
